reasoner/resolution/resolver/BoundConcludableResolver.java

```diff
@@ -0,0 +1,415 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.reasoner.resolution.resolver;
+
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.common.iterator.FunctionalIterator;
+import com.vaticle.typedb.core.common.iterator.Iterators;
+import com.vaticle.typedb.core.concept.ConceptManager;
+import com.vaticle.typedb.core.concept.answer.ConceptMap;
+import com.vaticle.typedb.core.logic.Rule;
+import com.vaticle.typedb.core.logic.resolvable.Concludable;
+import com.vaticle.typedb.core.logic.resolvable.Unifier;
+import com.vaticle.typedb.core.pattern.Conjunction;
+import com.vaticle.typedb.core.reasoner.resolution.ResolverRegistry;
+import com.vaticle.typedb.core.reasoner.resolution.answer.AnswerState;
+import com.vaticle.typedb.core.reasoner.resolution.framework.AnswerCache;
+import com.vaticle.typedb.core.reasoner.resolution.framework.AnswerCache.SubsumptionAnswerCache.ConceptMapCache;
+import com.vaticle.typedb.core.reasoner.resolution.framework.ReiterationQuery;
+import com.vaticle.typedb.core.reasoner.resolution.framework.Request;
+import com.vaticle.typedb.core.reasoner.resolution.framework.RequestState;
+import com.vaticle.typedb.core.reasoner.resolution.framework.Resolver;
+import com.vaticle.typedb.core.reasoner.resolution.framework.Response;
+import com.vaticle.typedb.core.traversal.TraversalEngine;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class BoundConcludableResolver extends Resolver<BoundConcludableResolver>  {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BoundConcludableResolver.class);
+    private AnswerCache<?, ConceptMap> cache;
+    private final Concludable concludable;
+    private final Set<Identifier.Variable.Retrievable> unboundVars;
+    private final ConceptMap bounds;
+    private final Map<Driver<ConclusionResolver>, Rule> resolverRules;
+    private final LinkedHashMap<Driver<ConclusionResolver>, Set<Unifier>> applicableRules;
+    private final Map<Request, RequestState.CachingRequestState<?, ConceptMap>> requestStates;
+    private RequestState exploringRequestState;
+
+    public BoundConcludableResolver(Driver<BoundConcludableResolver> driver, Concludable concludable, ConceptMap bounds,
+                                    Map<Driver<ConclusionResolver>, Rule> resolverRules,
+                                    LinkedHashMap<Driver<ConclusionResolver>, Set<Unifier>> applicableRules,
```

Comment:
any reason we are passing in applicable rules if we have it in the `Concludable`?

Reply:
the set of data structures is odd. The `concludable` can you get you applicable rules with unifiers. The `registry` can you `ConclusionResolvers` -- why do you need either of these maps (`resolverRules` and `applicableRules`?)

Reply:
This is a matter of whether we want to do `.initialiseDownstreamResolvers()` in the `ConcludableResolver` or `BoundConcludableResolver`. We can do it in the `BoundConcludableResolver` and rely on the fact that the `registry` does a put operation to create so that we don't have duplicate actors. For this decision I'm more inclined to think that it seems like the job of the coordinators to do this - if we move it to the bound resolvers then the coordinators become very thin classes.

Regarding  the `resolverRules` we can ask the registry for this if we store it there, or we can ask the resolver itself (not ideal) but if we're moving to have the coordinators spawn their workers then I'm not convinced. We don't want to be reading from the registry in that case right?

Reply:
i think what we'll end up with is an `initialiseDownstreamCoordinators` in each coordinator, which can be done with the registry. The registry will become a `coordinator` registry. 

This say we can use the registry and concludables to access the coordinators from the `Bound` resolvers. And we don't need all this state anymore :) 

Seem ok?

Reply:
Yes that sounds great!